File: examples/hscc16/franka/franka_two_agents.m
function franka
clear set
close all



%% simulation

% target set
lb1=[6 0];
ub1=lb1+0.5;
v1=[6 0; 6.5  0; 6 0.5; 6.5 .5];

lb2=[6 4];
ub2=lb2+0.5;
v2=[6 4; 6.5  4; 6 4.5; 6.5 4.5];
% initial state
x0_1=[0.4 0.4 0];
x0_2=[0.6 0.6 0];



controller1=SymbolicSet('franka_controller1.bdd','projection',[1 2 3]);
controller2=SymbolicSet('franka_controller2.bdd','projection',[1 2 3]);
target1=SymbolicSet('franka_target1.bdd');
target2=SymbolicSet('franka_target2.bdd');

y1=x0_1;
y2=x0_2;
v=[];
while(1)

  
  if ((target1.isElement(y1(end,:))) && (target2.isElement(y2(end,:))))
    break;
  end 

  u1=controller1.getInputs(y1(end,:));
  u2=controller2.getInputs(y2(end,:));

  v1=[v1; u1(1,:)];
  [t x1]=ode45(@unicycle_ode,[0 .3], y1(end,:), [],u1(1,:));

  y1=[y1; x1(end,:)];

  v2=[v2; u2(1,:)];
  [t x2]=ode45(@unicycle_ode,[0 .3], y2(end,:), [],u2(1,:));

  y2=[y2; x2(end,:)];
end



%% plot the franka domain
% colors
colors=get(groot,'DefaultAxesColorOrder');


% load the symbolic set containig the abstract state space
set=SymbolicSet('franka_ss.bdd','projection',[1 2]);
plotCells(set,'facecolor','none','edgec',[0.8 0.8 0.8],'linew',.1)
hold on

% load the symbolic set containig obstacles
set=SymbolicSet('franka_obst.bdd','projection',[1 2]);
plotCells(set,'facecolor',colors(1,:)*0.5+0.5,'edgec',colors(1,:),'linew',.1)

% plot the real obstacles and target set
plot_domain

% load the symbolic set containig target set
set=SymbolicSet('franka_target1.bdd','projection',[1 2]);
plotCells(set,'facecolor',colors(2,:)*0.5+0.5,'edgec',colors(2,:),'linew',.1)

set=SymbolicSet('franka_target2.bdd','projection',[1 2]);
plotCells(set,'facecolor',colors(2,:)*0.5+0.5,'edgec',colors(2,:),'linew',.1)

% plot initial state  and trajectory
plot(y1(:,1),y1(:,2),'k.-')
plot(y1(1,1),y1(1,1),'.','color',colors(5,:),'markersize',20)

plot(y2(:,1),y2(:,2),'k.-')
plot(y2(1,1),y2(1,1),'.','color',colors(5,:),'markersize',20)


box on
axis([-.5 10.5 -.5 10.5])


end

function dxdt = unicycle_ode(t,x,u)

  dxdt = zeros(3,1);
  c=atan(tan(u(2))/2);

  dxdt(1)=u(1)*cos(c+x(3))/cos(c);
  dxdt(2)=u(1)*sin(c+x(3))/cos(c);
  dxdt(3)=u(1)*tan(u(2));


end

function plot_domain

colors=get(groot,'DefaultAxesColorOrder');

v=[6 0; 6.5  0; 6 0.5; 6.5 .5];
patch('vertices',v,'faces',[1 2 4 3],'facec',colors(2,:),'edgec',colors(2,:));

v=[6 4; 6.5  4; 6 4.5; 6.5 4.5];
patch('vertices',v,'faces',[1 2 4 3],'facec',colors(2,:),'edgec',colors(2,:));

v=[2.2   0  ;2.4  0   ; 2.2   5    ; 2.4 5   ];
patch('vertices',v,'faces',[1 2 4 3],'facec',colors(1,:),'edgec',colors(1,:));
v=[4.6   1  ;4.8  1   ; 4.6   10   ; 4.8 10  ];
patch('vertices',v,'faces',[1 2 4 3],'facec',colors(1,:),'edgec',colors(1,:));

end

File: examples/hscc16/franka/franka_two_agents.cc
#include <array>
#include <iostream>

#include "cuddObj.hh"

#include "SymbolicSet.hh"
#include "SymbolicModelGrowthBound.hh"

#include "TicToc.hh"
#include "RungeKutta4.hh"
#include "FixedPoint.hh"

#ifndef M_PI
#define M_PI 3.14159265359
#endif


/* state space dim */
#define sDIM 3
#define iDIM 2

/* data types for the ode solver */
typedef std::array<double,3> state_type;
typedef std::array<double,2> input_type;

/* sampling time */
const double tau = 0.3;
/* number of intermediate steps in the ode solver */
const int nint=5;
OdeSolver ode_solver(sDIM,nint,tau);

/* we integrate the franka ode by 0.3 sec (the result is stored in x)  */
auto  franka_post = [](state_type &x, input_type &u) -> void {

  /* the ode describing the franka */
  auto rhs =[](state_type& xx,  const state_type &x, input_type &u) {
      double alpha=std::atan(std::tan(u[1])/2.0);
      xx[0] = u[0]*std::cos(alpha+x[2])/std::cos(alpha);
      xx[1] = u[0]*std::sin(alpha+x[2])/std::cos(alpha);
      xx[2] = u[0]*std::tan(u[1]);
  };
  ode_solver(rhs,x,u);
};

/* computation of the growth bound (the result is stored in r)  */
auto radius_post = [](state_type &r, input_type &u) {
    double c = std::abs(u[0]*std::sqrt(std::tan(u[1])*std::tan(u[1])/4.0+1));
    r[0] = r[0]+c*r[2]*0.3;
    r[1] = r[1]+c*r[2]*0.3;
};

/* forward declaration of the functions to setup the state space
 * input space and obstacles of the franka example */
scots::SymbolicSet frankaCreateStateSpace(Cudd &mgr);
scots::SymbolicSet frankaCreateInputSpace(Cudd &mgr);

void frankaCreateObstacles(scots::SymbolicSet &obs);


int main() {
  /* to measure time */
  TicToc tt;
  /* there is one unique manager to organize the bdd variables */
  Cudd mgr;

  /****************************************************************************/
  /* construct SymbolicSet for the state space */
  /****************************************************************************/
  scots::SymbolicSet ss=frankaCreateStateSpace(mgr);
  ss.writeToFile("franka_ss.bdd");

  /****************************************************************************/
  /* construct SymbolicSet for the obstacles */
  /****************************************************************************/
  /* first make a copy of the state space so that we obtain the grid
   * information in the new symbolic set */
  scots::SymbolicSet obs(ss);
  frankaCreateObstacles(obs);
  obs.writeToFile("franka_obst.bdd");

  /****************************************************************************/
  /* we define the target set */
  /****************************************************************************/
  /* first make a copy of the state space so that we obtain the grid
   * information in the new symbolic set */
  scots::SymbolicSet ts1(ss);
  /* define the target set as a symbolic set */
  double H[4*sDIM]={-1, 0, 0,
                    1, 0, 0,
                    0,-1, 0,
                    0, 1, 0};
  /* compute inner approximation of P={ x | H x<= h1 }  */
  double h1[4] = {-6,6.51,-0, .51};
  ts1.addPolytope(4,H,h1, scots::INNER);
  ts1.writeToFile("franka_target1.bdd");

  scots::SymbolicSet ts2(ss);
  double h2[4] = {-6,6.51,-4, 4.51};
  ts2.addPolytope(4,H,h2, scots::INNER);
  ts2.writeToFile("franka_target2.bdd");

  /****************************************************************************/
  /* construct SymbolicSet for the input space */
  /****************************************************************************/
  scots::SymbolicSet is=frankaCreateInputSpace(mgr);

  /****************************************************************************/
  /* setup class for symbolic model computation */
  /****************************************************************************/
  /* first create SymbolicSet of post variables 
   * by copying the SymbolicSet of the state space and assigning new BDD IDs */
  scots::SymbolicSet sspost(ss,1);
  /* instantiate the SymbolicModel */
  scots::SymbolicModelGrowthBound<state_type,input_type> abstraction(&ss, &is, &sspost);
  /* compute the transition relation */
  tt.tic();
  abstraction.computeTransitionRelation(franka_post, radius_post);
  std::cout << std::endl;
  tt.toc();
  /* get the number of elements in the transition relation */
  std::cout << std::endl << "Number of elements in the transition relation: " << abstraction.getSize() << std::endl;

  /****************************************************************************/
  /* we continue with the controller synthesis */
  /****************************************************************************/
  /* we setup a fixed point object to compute reachabilty controller */
  scots::FixedPoint fp1(&abstraction);
  /* the fixed point algorithm operates on the BDD directly */
  BDD T1 = ts1.getSymbolicSet();
  BDD O = obs.getSymbolicSet();
  tt.tic();
  /* compute controller */
  BDD C1=fp1.reachAvoid(T1,O,1);
  tt.toc();

  scots::FixedPoint fp2(&abstraction);
  /* the fixed point algorithm operates on the BDD directly */
  BDD T2 = ts2.getSymbolicSet();
  tt.tic();
  /* compute controller */
  BDD C2=fp2.reachAvoid(T2,O,1);
  tt.toc();

  /****************************************************************************/
  /* last we store the controller as a SymbolicSet 
   * the underlying uniform grid is given by the Cartesian product of 
   * the uniform gird of the space and uniform gird of the input space */
  /****************************************************************************/
  scots::SymbolicSet controller1(ss,is);
  controller1.setSymbolicSet(C1);
  controller1.writeToFile("franka_controller1.bdd");

  scots::SymbolicSet controller2(ss,is);
  controller2.setSymbolicSet(C2);
  controller2.writeToFile("franka_controller2.bdd");

  return 1;
}

scots::SymbolicSet frankaCreateStateSpace(Cudd &mgr) {

  /* setup the workspace of the synthesis problem and the uniform grid */
  /* lower bounds of the hyper rectangle */
  double lb[sDIM]={0,0,-M_PI-0.4};  
  /* upper bounds of the hyper rectangle */
  double ub[sDIM]={10,10,M_PI+0.4}; 
  /* grid node distance diameter */
  double eta[sDIM]={.2,.2,.2};   


  scots::SymbolicSet ss(mgr,sDIM,lb,ub,eta);

  /* add the grid points to the SymbolicSet ss */
  ss.addGridPoints();

  return ss;
}

void frankaCreateObstacles(scots::SymbolicSet &obs) {

  /* add the obstacles to the symbolic set */
  /* the obstacles are defined as polytopes */
  /* define H* x <= h */
  double H[4*sDIM]={-1, 0, 0,
                    1, 0, 0,
                    0,-1, 0,
                    0, 1, 0};

  double h2[4] = {-2.2,2.4,-0,5};
  obs.addPolytope(4,H,h2, scots::OUTER);
  double h5[4] = {-4.6 ,4.8,-1,10};
  obs.addPolytope(4,H,h5, scots::OUTER);

}

scots::SymbolicSet frankaCreateInputSpace(Cudd &mgr) {

  /* lower bounds of the hyper rectangle */
  double lb[sDIM]={-1,-1};  
  /* upper bounds of the hyper rectangle */
  double ub[sDIM]={1,1}; 
  /* grid node distance diameter */
  double eta[sDIM]={.3,.3};   

  scots::SymbolicSet is(mgr,iDIM,lb,ub,eta);
  is.addGridPoints();

  return is;
}